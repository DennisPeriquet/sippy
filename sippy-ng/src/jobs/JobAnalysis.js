import PropTypes from 'prop-types'
import React, { Fragment, useEffect } from 'react'
import { ArrayParam, JsonParam, useQueryParam } from 'use-query-params'
import Alert from '@material-ui/lab/Alert'
import { Box, Button, Card, Container, Dialog, Grid, Typography } from '@material-ui/core'
import './JobAnalysis.css'
import Divider from '@material-ui/core/Divider'
import SimpleBreadcrumbs from '../components/SimpleBreadcrumbs'
import { Link } from 'react-router-dom'
import { DataGrid } from '@material-ui/data-grid'
import { JobStackedChart } from './JobStackedChart'
import { Line } from 'react-chartjs-2'
import { scale } from 'chroma-js'
import GridToolbar from '../datagrid/GridToolbar'
import { explainFilter, pathForJobRunsWithFilter, pathForJobsWithFilter, withSort } from '../helpers'
import SummaryCard from '../components/SummaryCard'
import { JOB_THRESHOLDS } from '../constants'

export function JobAnalysis (props) {
  const [isLoaded, setLoaded] = React.useState(false)
  const [analysis, setAnalysis] = React.useState({ by_day: {} })
  const [filterModel] = useQueryParam('filters', JsonParam)

  const [fetchError, setFetchError] = React.useState('')

  const [allTests, setAllTests] = React.useState([])
  const [selectionModel, setSelectionModel] = React.useState([0, 1, 2, 3, 4])
  const [selectedTests = [], setSelectedTests] = useQueryParam('tests', ArrayParam)
  const [testFilter = { items: [] }, setTestFilter] = useQueryParam('testFilters', JsonParam)
  const [testSelectionDialog, setTestSelectionDialog] = React.useState(false)

  const fetchData = () => {
    document.title = `Sippy > ${props.release} > Jobs > Analysis`

    let queryParams = `release=${props.release}`
    if (filterModel) {
      queryParams += `&filter=${encodeURIComponent(JSON.stringify(filterModel))}`
    }

    Promise.all([
      fetch(`${process.env.REACT_APP_API_URL}/api/jobs/analysis?${queryParams}`)
    ])
      .then(([analysis]) => {
        if (analysis.status !== 200) {
          throw new Error('server returned ' + analysis.status)
        }

        return Promise.all([analysis.json()])
      })
      .then(([analysis]) => {
        setAnalysis(analysis)

        let allTests = new Map()
        Object.keys(analysis.by_day)
          .map((key) => Object.keys(analysis.by_day[key].test_count)
            .forEach((test) => {
              const count = allTests.get(test) || { name: test, value: 0 }
              allTests.set(test, { name: test, value: count.value + analysis.by_day[key].test_count[test] })
            }))

        allTests = [...allTests.values()].sort((a, b) => b.value - a.value)

        const selected = []
        // DataGrid tables need an ID:
        for (let i = 0; i < allTests.length; i++) {
          if (selectedTests && selectedTests.includes(allTests[i].name)) {
            selected.push(i)
          }
          allTests[i].id = i
        }

        if (selected.length > 0) {
          setSelectionModel(selected)
        }

        setAllTests(allTests)
        setLoaded(true)
      }
      ).catch(error => {
        setFetchError('Could not retrieve job analysis ' + props.release + ', ' + error)
      })
  }

  useEffect(() => {
    fetchData()
  }, [])

  if (fetchError !== '') {
    return <Alert severity="error">{fetchError}</Alert>
  }

  if (!isLoaded) {
    return (
      <p>Loading...</p>
    )
  }

  const topTestChart = {
    labels: Object.keys(analysis.by_day),
    datasets: []
  }

  const colors = scale('Set2').mode('lch').colors(selectionModel.length)

  selectionModel.forEach((id, index) => {
    if (id < allTests.length) {
      topTestChart.datasets.push({
        type: 'line',
        label: allTests[id].name,
        backgroundColor: colors[index],
        borderColor: colors[index],
        tension: 0.3,
        data: Object.keys(analysis.by_day).map((key) => (100 * (1 - (analysis.by_day[key].test_count[allTests[id].name] || 0) / analysis.by_day[key].total_runs)))
      })
    }
  })

  const options = {
    plugins: {
      tooltip: {
        callbacks: {
          label: function (context) {
            const failures = analysis.by_day[context.label].test_count[context.dataset.label] || 0
            const runs = analysis.by_day[context.label].total_runs

            return `${context.dataset.label} ${context.raw.toFixed(2)}% (${failures} failed of ${runs} runs)`
          }
        }
      }
    },
    scales: {
      y: {
        max: 100,
        ticks: {
          callback: (value, index, values) => {
            return `${value}%`
          }
        }
      }
    }
  }

  const requestSearch = (searchValue) => {
    const currentFilters = testFilter
    currentFilters.items = currentFilters.items.filter(
      (f) => f.columnField !== 'name'
    )
    currentFilters.items.push({
      id: 99,
      columnField: 'name',
      operatorValue: 'contains',
      value: searchValue
    })
    setTestFilter(currentFilters)
  }

  const updateSelectionModel = (m) => {
    setSelectedTests(allTests.filter((row) => m.includes(row.id)).map((row) => row.name))
    setSelectionModel(m)
  }

  const totalSuccess = Object.keys(analysis.by_day)
    .map((key) => analysis.by_day[key].result_count.S || 0)
    .reduce((acc, val) => acc + val)

  const totalRuns = Object.keys(analysis.by_day)
    .map((key) => analysis.by_day[key].total_runs)
    .reduce((acc, val) => acc + val)

  return (
    <Fragment>
      <SimpleBreadcrumbs
        release={props.release}
        previousPage={
          <Link
            to={pathForJobsWithFilter(props.release, filterModel)}>
            Jobs
          </Link>
        }
        currentPage="Job Analysis"
      />

      <Container size="xl">
        <Typography variant="h3" style={{ textAlign: 'center' }}>
          Job Analysis
        </Typography>
        <Divider style={{ margin: 20 }} />

        <Grid container spacing={3}>
          <Grid item md={4}>
            <SummaryCard
              key="test-summary"
              threshold={JOB_THRESHOLDS}
              name="Overall"
              success={totalSuccess}
              fail={totalRuns - totalSuccess}
              flakes={0}
              caption={
                <Fragment>{totalRuns} runs, {totalSuccess} successful</Fragment>
              }
            />
          </Grid>

          <Grid item md={8}>
            <Card className="test-failure-card" elevation={5} style={{ height: '100%' }}>
              <strong>Current filter</strong>:<br />
              Showing jobs matching {explainFilter(filterModel)}<br />
              <Divider />
              <Button
                variant="contained"
                color="primary"
                component={Link}
                style={{ marginTop: 20, marginRight: 20 }}
                to={pathForJobsWithFilter(props.release, filterModel)}
              >
                View matching jobs
              </Button>

              <Button
                variant="contained"
                color="secondary"
                component={Link}
                style={{ marginTop: 20 }}
                to={withSort(pathForJobRunsWithFilter(props.release, filterModel), 'timestamp', 'desc')}
              >
                View matching job runs
              </Button>
            </Card>
          </Grid>

          <Grid item md={12}>
            <Card className="job-failure-card" elevation={5}>
              <Typography variant="h5">
                Job results
              </Typography>
              <JobStackedChart release={props.release} analysis={analysis} />
            </Card>
          </Grid>

          <Grid item md={12}>
            <Card className="job-failure-card" elevation={5}>
              <Typography variant="h5">
                Test results for selected jobs
              </Typography>
              <Box hidden={allTests.length !== 0}>
                <p>No test data found.</p>
              </Box>
              <Box hidden={allTests.length === 0}>
                <Box hidden={selectionModel.length === 0}>
                  <Line data={topTestChart} options={options} height={100} />
                </Box>
                <Button style={{ marginTop: 20 }} variant="contained" color="secondary" onClick={() => setTestSelectionDialog(true)}>Select tests to chart</Button>
                <Dialog fullWidth={true} maxWidth="lg" onClose={() => setTestSelectionDialog(false)} open={testSelectionDialog}>
                  <Grid className="test-dialog">
                    <Typography variant="h6" style={{ marginTop: 20, marginBottom: 20 }}>
                      Select tests to chart
                    </Typography>
                    <DataGrid
                      components={{ Toolbar: GridToolbar }}
                      columns={
                        [
                          {
                            field: 'id',
                            hide: true,
                            filterable: false
                          },
                          {
                            field: 'name',
                            headerName: 'Test name',
                            flex: 4,
                            renderCell: (param) => (
                              <div className="job-name">
                                {param.value}
                              </div>
                            )
                          },
                          {
                            field: 'value',
                            type: 'number',
                            headerName: 'Failure count',
                            flex: 1
                          }
                        ]
                      }
                      rows={allTests}
                      pageSize={10}
                      rowHeight={60}
                      autoHeight={true}
                      selectionModel={selectionModel}
                      onSelectionModelChange={(m) => updateSelectionModel(m)}
                      filterModel={testFilter}
                      onFilterModelChange={(m) => { setTestFilter(m) }}
                      checkboxSelection
                      componentsProps={{
                        toolbar: {
                          setFilterModel: setTestFilter,
                          clearSearch: () => requestSearch(''),
                          doSearch: requestSearch
                        }
                      }}
                    />

                    <Button style={{ marginTop: 20 }} variant="contained" color="primary" onClick={() => setTestSelectionDialog(false)}>OK</Button>
                  </Grid>
                </Dialog>
              </Box>
            </Card>
          </Grid>
        </Grid>
      </Container>
    </Fragment>
  )
}

JobAnalysis.defaultProps = {
  job: ''
}

JobAnalysis.propTypes = {
  release: PropTypes.string.isRequired,
  job: PropTypes.string
}
