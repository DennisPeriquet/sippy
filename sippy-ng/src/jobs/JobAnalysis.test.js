/** @jest-environment setup-polly-jest/jest-environment-node */

import 'jsdom-global/register'
import { act, waitFor } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import {
  expectLoadingPage,
  setupDefaultPolly,
  withoutMuiID,
} from '../setupTests'
import { JobAnalysis } from './JobAnalysis'
import { mount } from 'enzyme'
import { QueryParamProvider } from 'use-query-params'
import React from 'react'

jest.useRealTimers()

describe('JobAnalysis', () => {
  setupDefaultPolly()

  beforeEach(() => {
    Date.now = jest
      .spyOn(Date, 'now')
      .mockImplementation(() => new Date(1628691480000))
  })

  it('should render correctly', async () => {
    const fetchSpy = jest.spyOn(global, 'fetch')

    let wrapper
    await act(async () => {
      wrapper = mount(
        <QueryParamProvider>
          <BrowserRouter>
            <JobAnalysis release="4.8" />
          </BrowserRouter>
        </QueryParamProvider>
      )
    })

    expectLoadingPage(wrapper).toBeTruthy()

    await waitFor(() => {
      wrapper.update()
      expectLoadingPage(wrapper).toBeFalsy()
    })

    expect(wrapper.text()).toContain('Job results')
    expect(wrapper.exists()).toBe(true)
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
    expect(fetchSpy).toHaveBeenCalledTimes(1)
  })
})
