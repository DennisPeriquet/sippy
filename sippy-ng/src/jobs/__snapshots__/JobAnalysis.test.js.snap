// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobAnalysis should render correctly 1`] = `
<nav class="MuiTypography-root MuiBreadcrumbs-root MuiTypography-body1 MuiTypography-colorTextSecondary"
     aria-label="breadcrumb"
>
  <ol class="MuiBreadcrumbs-ol">
    <li class="MuiBreadcrumbs-li">
      <a href="/release/4.8">
        Overview
      </a>
    </li>
    <li aria-hidden="true"
        class="MuiBreadcrumbs-separator"
    >
      /
    </li>
    <li class="MuiBreadcrumbs-li">
      <a href="/jobs/4.8">
        Jobs
      </a>
    </li>
    <li aria-hidden="true"
        class="MuiBreadcrumbs-separator"
    >
      /
    </li>
    <li class="MuiBreadcrumbs-li">
      <p class="MuiTypography-root MuiTypography-body1">
        Job Analysis
      </p>
    </li>
  </ol>
</nav>
<div class="MuiContainer-root MuiContainer-maxWidthLg">
  <h3 class="MuiTypography-root MuiTypography-h3"
      style="text-align: center;"
  >
    Job Analysis
  </h3>
  <hr class="MuiDivider-root"
      style="margin: 20px;"
  >
  <div class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3">
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4">
      <div class="MuiPaper-root MuiCard-root makeStyles-summaryCard-2 makeStyles-summaryCard-3 MuiPaper-elevation5 MuiPaper-rounded"
           style="background-color: rgb(129, 199, 132);"
           title
      >
        <div class="MuiCardContent-root makeStyles-cardContent-1">
          <h6 class="MuiTypography-root MuiTypography-h6">
            Overall
            <svg class="MuiSvgIcon-root"
                 focusable="false"
                 viewbox="0 0 24 24"
                 aria-hidden="true"
            >
              <path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z">
              </path>
            </svg>
          </h6>
          <div align="center">
            <div style="width: 70%;">
            </div>
            5788 runs, 4677 successful
          </div>
        </div>
      </div>
    </div>
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-8">
      <div class="MuiPaper-root MuiCard-root test-failure-card MuiPaper-elevation5 MuiPaper-rounded"
           style="height: 100%;"
      >
        <strong>
          Current filter
        </strong>
        <br>
        Showing all results
        <br>
        <hr class="MuiDivider-root"
            style="margin-bottom: 20px;"
        >
        <div class="MuiGrid-root MuiGrid-container MuiGrid-justify-content-xs-space-between">
          <button class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                  tabindex="0"
                  type="button"
                  title="Showing all results"
          >
            <span class="MuiButton-label">
              <span class="MuiBadge-root">
                <svg class="MuiSvgIcon-root"
                     focusable="false"
                     viewbox="0 0 24 24"
                     aria-hidden="true"
                >
                  <path d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z">
                  </path>
                </svg>
                <span class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary MuiBadge-invisible">
                  0
                </span>
              </span>
              Filters
            </span>
            <span class="MuiTouchRipple-root">
            </span>
          </button>
          <button class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                  tabindex="0"
                  type="button"
          >
            <span class="MuiButton-label">
              View by hour
            </span>
            <span class="MuiTouchRipple-root">
            </span>
          </button>
          <a class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
             tabindex="0"
             role="button"
             aria-disabled="false"
             style="margin-left: 20px; margin-right: 20px;"
             href="/jobs/4.8"
          >
            <span class="MuiButton-label">
              View matching jobs
            </span>
            <span class="MuiTouchRipple-root">
            </span>
          </a>
          <a class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
             tabindex="0"
             role="button"
             aria-disabled="false"
             href="/jobs/4.8/runs?sortField=timestamp&amp;sort=desc"
          >
            <span class="MuiButton-label">
              View matching job runs
            </span>
            <span class="MuiTouchRipple-root">
            </span>
          </a>
        </div>
      </div>
    </div>
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-12">
      <div class="MuiPaper-root MuiCard-root job-failure-card MuiPaper-elevation5 MuiPaper-rounded">
        <h5 class="MuiTypography-root MuiTypography-h5">
          Job results
        </h5>
      </div>
    </div>
    <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-12">
      <div class="MuiPaper-root MuiCard-root job-failure-card MuiPaper-elevation5 MuiPaper-rounded">
        <h5 class="MuiTypography-root MuiTypography-h5">
          Test results for selected jobs
          <svg class="MuiSvgIcon-root"
               focusable="false"
               viewbox="0 0 24 24"
               aria-hidden="true"
               title="By default, this plots the top 5 tests with the most failures in the set of jobs selected. Click the button below to change the list of tests to plot. Only tests with at least one failure during the reporting period are available."
          >
            <path d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z">
            </path>
          </svg>
        </h5>
        <div class="MuiBox-root MuiBox-root-8"
             hidden
        >
          <p>
            No test data found.
          </p>
        </div>
        <div class="MuiBox-root MuiBox-root-9">
          <div class="MuiBox-root MuiBox-root-10">
          </div>
          <button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
                  tabindex="0"
                  type="button"
                  style="margin-top: 20px;"
          >
            <span class="MuiButton-label">
              Select tests to chart
            </span>
            <span class="MuiTouchRipple-root">
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
